File: Bronze/2016/February/01_milk_pails/MilkPails2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class MilkPails2 {
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("pails.in"));
        PrintWriter pw = new PrintWriter("pails.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int maxX = m / x;
        // int maxY = m / y;
        int ans = 0;
        // 搜索范围内遍历得到最大值
        for (int i = 0; i <= maxX; i++) {
            // 对于给定的 i 个 x 确定 y 的数量为 (m - x * i) / y
            int j = (m - x * i) / y;
            ans = Math.max(ans, x * i + y * j);
        }
        pw.println(ans);
        pw.close();
    }
}

File: Bronze/2016/February/01_milk_pails/milk_pails2.cpp
#include <iostream>
#include <fstream>
#include <algorithm>

using namespace std;

int main() {
    freopen("pails.in", "r", stdin);
    freopen("pails.out", "w", stdout);

    int X, Y, M;
    cin >> X >> Y >> M;

    int maxX = M / X;
    int maxY = M / Y;
    int ans = 0;

    // 搜索范围内遍历得到最大值
    for (int i = 0; i <= maxX; i++) {
        for (int j = 0; j <= maxY; j++) {
            if (X * i + Y * j <= M) {
                ans = max(ans, X * i + Y * j);
            }
        }
    }

    cout << ans << endl;

    return 0;
}
